src/truss.cpp

```diff
@@ -0,0 +1,242 @@
+
+/**
+ * \function trussness
+ * \brief Find the "trussness" for every edge in the graph. This
+ * indicates the highest k-truss that the edge occurs in.
+ *
+ * \param graph The input graph.
+ * \param truss Pointer to initialized vector of truss values that will
+ * indicate the highest k-truss each edge occurs in. It will be resized as
+ * needed.
+ *
+ * \example examples/simple/igraph_truss.cpp
+ */
+int igraph_truss(const igraph_t* graph, igraph_vector_int_t* truss){
+  igraph_vector_int_t triangles, support;
+  igraph_vector_t eid, unpacked_triangles;
+
+  // List the triangles as vertex triplets.
+  igraph_vector_int_init(&triangles, 0);  // will be resized
+  igraph_list_triangles(graph, &triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &triangles);
+
+  // Unpack the triangles from vertex list to edge list.
+  long int num_triangles = igraph_vector_int_size(&triangles);
+  igraph_vector_init(&unpacked_triangles, 2 * num_triangles);
+  unpack(&triangles, &unpacked_triangles);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &unpacked_triangles);
+  igraph_vector_int_destroy(&triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Get the edge ids of the unpacked triangles. Note: a given eid can occur
+  // multiple times in this list if it is in multiple triangles.
+  igraph_vector_init(&eid, num_triangles);
+  igraph_get_eids(graph, &eid, &unpacked_triangles, 0, 0, 1);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &eid);
+  igraph_vector_destroy(&unpacked_triangles);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the support of the edges.
+  igraph_vector_int_init(&support, igraph_ecount(graph));
+  compute_support(&eid, &support);
+  IGRAPH_FINALLY(igraph_vector_int_destroy, &support);
+  igraph_vector_destroy(&eid);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  // Compute the truss of the edges.
+  igraph_vector_int_init(truss, igraph_ecount(graph));
+  trussness(graph, &support, truss);
+  igraph_vector_int_destroy(&support);
+  IGRAPH_FINALLY_CLEAN(1);
+
+  return 0;
+}
+
+// Unpack the triangles as a vector of vertices to be a vector of edges.
+// So, instead of the triangle specified as vertices [1, 2, 3], return the
+// edges as [1, 2, 1, 3, 2, 3] so that the support can be computed.
+void unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri) {
+  int j;
+  for (long int i = 0; i < igraph_vector_int_size(tri); i = i + 3){
+    j = i * 2;
+    VECTOR(*unpacked_tri)[j]   = VECTOR(*unpacked_tri)[j+2] = VECTOR(*tri)[i];
+    VECTOR(*unpacked_tri)[j+1] = VECTOR(*unpacked_tri)[j+4] = VECTOR(*tri)[i+1];
+    VECTOR(*unpacked_tri)[j+3] = VECTOR(*unpacked_tri)[j+5] = VECTOR(*tri)[i+2];
+  }
+}
+
+// Compute the edge support, i.e. number of triangles each edge occurs in.
+// Time complexity: O(m), where m is the number of edges listed in eid.
+void compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support) {
+  for (long int i = 0; i < igraph_vector_size(eid); ++i){
+    VECTOR(*support)[(int) VECTOR(*eid)[i]] += 1;
```

Comment:
Casting to `long int` here would probably be safer - `int` might be only 16 bits on some (very old) systems.